### PR TITLE
Mamba/perf v1

### PR DESCRIPTION
**Pull request by kpaigwar**
* Added optimized models and their tests
* Reshape, permute, repeat_interleave are still fallbacks
* Added support for tensor caching
* Added both functional and optimized unit tests in CI

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

File: tests/scripts/run_models.sh [hunk 1]
```diff
@@ -15,6 +15,8 @@ fi
 if [[ $ARCH_NAME == "wormhole" ]]; then
   env pytest tests/ttnn/integration_tests/stable_diffusion
   env pytest tests/ttnn/integration_tests/unet
+  env pytest models/experimental/mamba/tests
```
Comment (tt-rkim) [inline tests/scripts/run_models.sh:18]: Hey @kpaigwar 

So I would actually recommend putting in a different if block with the proper arch name. There is actually a typo here - it should be `wormhole_b0` and not `wormhole`. I believe the original authors of this if block did not check this.

I would also recommend putting in a `@skip_for_grayskull` decorator with a reason in English to be sure that people see what's going on when they skip it on GS.

```
if [[ $ARCH_NAME == "wormhole_b0" ]]; then
  env pytest models/experimental/mamba/tests
fi
```

here. And then you should run the fast dispatch nightly tests: https://github.com/tenstorrent-metal/tt-metal/actions/workflows/full-regressions-and-models.yaml

Verify that it is running on CI and paste the passing link in this PR.

It should be one of the first models run, so you should have pretty quick feedback.

**Review by tt-rkim — CHANGES_REQUESTED** [review]

File: tests/scripts/run_models.sh (hunk 1 above)
Comment (tt-rkim) [reply tests/scripts/run_models.sh:18]: We are currently working on skipping any failing test in that pipeline, but you should not be blocked from doing your work and your mamba test should be one of the first run on `wormhole_b0, models` variant of fast dispatch pipeline.

Comment (kpaigwar) [reply tests/scripts/run_models.sh]: Hey @tt-rkim, thanks for the review. I added suggested changes. And here is the link for fast-dispatch nightly run https://github.com/tenstorrent-metal/tt-metal/actions/runs/8461855598

**Comment by tt-rkim** [comment]
Looks great, just a couple more questions:

- When is the README coming for the demo?
- When is the perf test coming?
- When do you see being able to add mamba to the 2x4 T3000 CI?

**Comment by kpaigwar** [comment]
@tt-rkim 
I updated the README with instructions to run our functional demos.
1. [Perf Report] We are still in non-performant stage where some of our Ops are Host fallbacks. We are working on the performant version. PR will be coming soon.
2. [2x4 T3000 CI] Mamba is small model which can easily run on a single WH. Data-parallel version is on-going.

**Review by tt-rkim — APPROVED** [review]
Sounds good, I see model passing in nightly CI:

```
2024-03-28T03:21:51.1118385Z [38;2;000;128;000m                  Metal[0m | [1m[38;2;100;149;237mINFO    [0m | Initializing device 0
2024-03-28T03:21:51.1647850Z [38;2;000;128;000m                  Metal[0m | [1m[38;2;100;149;237mINFO    [0m | AI CLK for device 0 is:   800 MHz
2024-03-28T03:23:39.4928363Z Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
2024-03-28T03:25:20.0047570Z 2024-03-28 03:25:20.004 | INFO     | models.experimental.mamba.tests_opt.test_full_model:test_mamba_model_inference:75 - (False, 'Max ATOL Delta: 3.0277438163757324, Max RTOL Delta: 8591.0498046875, PCC: 0.9957704284631479, Allclose check failed')
2024-03-28T03:25:20.0484945Z 2024-03-28 03:25:20.047 | INFO     | models.experimental.mamba.tests_opt.test_full_model:test_mamba_model_inference:78 - PCC value: Max ATOL Delta: 3.0277438163757324, Max RTOL Delta: 8591.0498046875, PCC: 0.9957704284631479
2024-03-28T03:25:20.3805929Z Initalizing MambaTT with 64 layers
2024-03-28T03:25:20.3811587Z PASSED[38;2;000;128;000m                  Metal[0m | [1m[38;2;100;149;237mINFO    [0m | Closing device 0
2024-03-28T03:25:20.4211324Z 
2024-03-28T03:25:20.4211647Z 
2024-03-28T03:25:20.4212384Z ==================================== PASSES ====================================
2024-03-28T03:25:20.4213085Z =========================== short test summary info ============================
2024-03-28T03:25:20.4220573Z PASSED models/experimental/mamba/tests_opt/test_full_model.py::test_mamba_model_inference[state-spaces/mamba-2.8b-32-0.99-False]
```

Great work, thank you

**Pull request merged**